tracer/src/Datadog.Trace.Manual/Configuration/ImmutableTracerSettings.cs

```diff
@@ -0,0 +1,191 @@
+﻿// <copyright file="ImmutableTracerSettings.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System.Collections.Concurrent;
+using Datadog.Trace.ClrProfiler.AutoInstrumentation.ManualInstrumentation;
+using Datadog.Trace.SourceGenerators;
+
+namespace Datadog.Trace.Configuration;
+
+/// <summary>
+/// Contains Tracer settings.
+/// </summary>
+public sealed class ImmutableTracerSettings
+{
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ImmutableTracerSettings"/> class.
+    /// </summary>
+    internal ImmutableTracerSettings(Dictionary<string, object?> initialValues)
+    {
+        AgentUri = GetValue<Uri?>(initialValues, TracerSettingKeyConstants.AgentUriKey, null) ?? new Uri("http://127.0.0.1:8126");
+        CustomSamplingRules = GetValue<string?>(initialValues, TracerSettingKeyConstants.CustomSamplingRules, null);
+        Environment = GetValue<string?>(initialValues, TracerSettingKeyConstants.EnvironmentKey, null);
+        GlobalSamplingRate = GetValue<double?>(initialValues, TracerSettingKeyConstants.GlobalSamplingRateKey, null);
+        GlobalTags = GetValue<IReadOnlyDictionary<string, string>?>(initialValues, TracerSettingKeyConstants.GlobalTagsKey, null) ?? new ConcurrentDictionary<string, string>();
```

Comment:
NIT: being read only is it really needed it to be thread safe?

Reply:
Yes and no 😄 This is essentially a historic decision. Previously, `TracerSettings` and `ImmutableTracerSettings` were one and the same, so customers _could_ modify it any time. When we split them, we kept both returning the same types for back-compat reasons. This is essentially duplicating the same behaviour.

That said, I agree, I think now is probably the time to use a "real" readonly dictionary/singleton here - if not, then when?! 😄

Reply:
FWIW, those fallback values are only created if the customer is using manual instrumentation _without_ automatic instrumentation (i.e. a no-op mode we don't support). So it's probably not worth the hassle tbh 😅

Reply:
Decided not to address this one as this is very niche, and only occurs in manual-only instrumentation